torch/_export/__init__.py

```diff
@@ -55,6 +55,499 @@
+    """
+    # Get parameter and buffer dictionaries from graph module
+    params_dict = dict(gm.named_parameters())
+    buffers_dict = dict(gm.named_buffers())
+
+    # Look up params and buffers by FQN in the order specified by specs
+    params = [params_dict[fqn] for fqn in params_spec]
+    buffers = [buffers_dict[fqn] for fqn in buffers_spec]
+
+    def wrapper(*args, **kwargs):
+        # Flatten the inputs using in_spec to handle structured inputs (dicts, tuples, etc.)
+        # The in_spec includes params/buffers/inputs, so we reconstruct the full input structure
+        # and flatten just the user inputs portion
+        user_inputs_flat, _ = pytree.tree_flatten((args, kwargs))
+        # Construct the full flat args list
+        flat_args = [*params, *buffers, *user_inputs_flat]
+        # Call the compiled function
+        flat_outputs = compiled_fn(flat_args)
+        # Unflatten outputs using out_spec to handle structured outputs
+        return pytree.tree_unflatten(flat_outputs, out_spec)
+
+    return wrapper
+
+
+class CompilerPipeline:
+    """
+    A unified pipeline for graph capture, joint graph generation, and compilation.
+
+    This class provides an end-to-end API for:
+    1. Capturing a graph from an nn.Module using Dynamo
+    2. Generating a joint forward-backward graph with descriptors
+    3. Partitioning the joint graph into forward and backward graphs
+    4. Compiling forward and backward graphs
+    5. Creating autograd or inference functions
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        inputs: tuple[Any, ...],
+        kwargs: Optional[dict[str, Any]] = None,
+    ):
+        """
+        Initialize the CompilerPipeline.
+
+        Args:
+            model: The nn.Module to capture and compile
+            inputs: Example input tensors for tracing
+            kwargs: Optional keyword arguments for the model
+
+        NOTE: You must use the following pattern to use this class:
+
+        pipeline = CompilerPipeline(...)  # capture graph
+        with pipeline.stack:
+            # call pipeline methods to compile
+        """
+        self.model = model
+        self.inputs = inputs
+        self.kwargs = kwargs if kwargs is not None else {}
+        self._gm = self.capture_graph()
+
+        self.stack = ExitStack()
+
+        self.joint_with_descriptors: Optional[Any] = None  # JointWithDescriptors
+
+        # Internal state for compilation phases
+        self._aot_config: Optional[Any] = None
+        self._fw_metadata: Optional[Any] = None
+        self._maybe_subclass_meta: Optional[Any] = None
+
+    def capture_graph(self) -> torch.fx.GraphModule:
+        """
+        Capture the graph from the model using Dynamo graph capture.
+
+        Returns:
+            The captured GraphModule
+        """
+        from torch._dynamo.functional_export import _dynamo_graph_capture_for_export, _restore_state_dict
```

Comment:
I think _dynamo_graph_capture_for_export is outdated now and we should just use `dynamo_graph_capture_for_export` w/o _restore_state_dict fine.